test/fuzzers/random_generator.cpp

```diff
@@ -0,0 +1,133 @@
+  return std::uniform_int_distribution<I>(lower, upper - 1)(*engine);
+}
+
+/// Helper for obtaining a seed bias value for HashCombine with a bit-width
+/// dependent on the size of size_t.
+template <int SIZE_OF_SIZE_T>
+struct HashCombineOffset {};
+/// Specialization of HashCombineOffset for size_t == 4.
+template <>
+struct HashCombineOffset<4> {
+  /// @returns the seed bias value for HashCombine()
+  static constexpr inline uint32_t value() {
+    return 0x9e3779b9;  // Fractional portion of Golden Ratio, suggested by
+                        // Linux Kernel and Knuth's Art of Computer Programming
+  }
+};
+/// Specialization of HashCombineOffset for size_t == 8.
+template <>
+struct HashCombineOffset<8> {
+  /// @returns the seed bias value for HashCombine()
+  static constexpr inline uint64_t value() {
+    return 0x9e3779b97f4a7c16;  // Fractional portion of Golden Ratio, suggested
+                                // by Linux Kernel and Knuth's Art of Computer
+                                // Programming
+  }
+};
+
+/// HashCombine "hashes" together an existing hash and hashable values.
+template <typename T>
+void HashCombine(size_t* hash, const T& value) {
+  constexpr size_t offset = HashCombineOffset<sizeof(size_t)>::value();
+  *hash ^= std::hash<T>()(value) + offset + (*hash << 6) + (*hash >> 2);
+}
+
+/// Calculate the hash for the contents of a C-style data buffer
+/// @param data - pointer to buffer to be hashed
+/// @param size - number of elements in buffer
+/// @returns hash of the data in the buffer
+size_t HashBuffer(const uint8_t* data, const size_t size) {
+  size_t hash = 0xCA8945571519E991;  // seed with an arbitrary prime
+  HashCombine(&hash, size);
+  for (size_t i = 0; i < size; i++) {
+    HashCombine(&hash, data[i]);
+  }
+  return hash;
+}
+
+}  // namespace
+
+RandomGenerator::RandomGenerator(uint64_t seed) : engine_(seed) {}
+
+RandomGenerator::RandomGenerator(const uint8_t* data, size_t size) {
+  RandomGenerator(RandomGenerator::CalculateSeed(data, size));
+}
+
+spv_target_env RandomGenerator::GetTargetEnv() {
+  // SPV_ENV_WEBGPU_0 is intentionally omitted here, since it is deprecated and
+  // using it will cause asserts.
+  static const std::array<spv_target_env, 23> envs = {
+      SPV_ENV_UNIVERSAL_1_0,       SPV_ENV_VULKAN_1_0,
+      SPV_ENV_UNIVERSAL_1_1,       SPV_ENV_UNIVERSAL_1_2,
+      SPV_ENV_UNIVERSAL_1_3,       SPV_ENV_VULKAN_1_1,
+      SPV_ENV_OPENCL_1_2,          SPV_ENV_OPENCL_EMBEDDED_1_2,
+      SPV_ENV_OPENCL_2_0,          SPV_ENV_OPENCL_EMBEDDED_2_0,
+      SPV_ENV_OPENCL_EMBEDDED_2_1, SPV_ENV_OPENCL_EMBEDDED_2_2,
+      SPV_ENV_OPENCL_2_1,          SPV_ENV_OPENCL_2_2,
+      SPV_ENV_UNIVERSAL_1_4,       SPV_ENV_VULKAN_1_1_SPIRV_1_4,
+      SPV_ENV_UNIVERSAL_1_5,       SPV_ENV_VULKAN_1_2,
+      SPV_ENV_OPENGL_4_0,          SPV_ENV_OPENGL_4_1,
+      SPV_ENV_OPENGL_4_2,          SPV_ENV_OPENGL_4_3,
+      SPV_ENV_OPENGL_4_5};
```

Comment:
My problem with this is that it requires maintenance: as more versions are added, which they will be over time, someone needs to remember to update this.

If we could add a SPV_ENV_MAX value to the enum this problem would go away.

I see your point about testing code not necessarily being sufficient motivation for this. But actually OSS-Fuzz has found an undefined beahvior in spirv-val related to this problem, where an out of range value is cast to this enum type. With no easy way to know how large the enum type is, it's hard to fix the UB.

So: I'm fine with this for now but I think we should discuss it - I have opened https://github.com/KhronosGroup/SPIRV-Tools/issues/4558.